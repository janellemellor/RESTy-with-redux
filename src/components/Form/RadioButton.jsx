import React from 'react';
import PropTypes from 'prop-types';

export const RadioButtonGroup = ({ children, name, onChange }) => {
  const radioButtons = React.Children.map(children, child => {
    return React.cloneElement(child, {
      name, 
      onChange
    });
  });

  return (
    <>
      {radioButtons}
    </>
  );
};

export const RadioButton = ({ name, value, onChange }) => (
  <>
    <input type ="radio" id={value} name={name} value={value} onChange={onChange} />
    <label htmlFor={value}>{value}</label>
  </>
);

RadioButtonGroup.propTypes = {
  children: PropTypes.node, 
  name: PropTypes.string.isRequired, 
  onChange: PropTypes.func.isRequired
};

RadioButton.propTypes = {
  name: PropTypes.string.isRequired, 
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string
};
