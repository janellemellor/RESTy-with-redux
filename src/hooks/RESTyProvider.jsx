import React, { createContext, useState, useContext } from 'react';

const RESTyContext = createContext();

export const RESTyProvider = ({ children }) => {
  const [url, setUrl] = useState('');
  const [method, setMethod] = useState('GET');
  const [jsonBody, setJsonBody] = useState('');
  
  const onChange = ({ target }) => {
    const setInputsFactory = {
      url: setUrl, 
      method: setMethod, 
      jsonBody: setJsonBody
    };
    setInputsFactory[target.name](target.value);
  };

  const context = {
    url,
    method, 
    jsonBody, 
    onChange
  };

  return (
    <RESTyContext.Provider value={context} >
      {children}
    </RESTyContext.Provider>
  );
};

export const useRESTy = () => {
  const context = useContext(RESTyContext);
  return context;
};
