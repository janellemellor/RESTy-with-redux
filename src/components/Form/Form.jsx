import React from 'react';
import { RadioButtonGroup, RadioButton } from './RadioButton';
import { useRESTy } from '../../hooks/RESTyProvider';

const Form = () => {
  const { url, onChange, jsonBody } = useRESTy();

  return (
    <form onSubmit={onSubmit}>
      <fieldset>
        <input name="url" type="text" value={url} onChange={onChange} />
        <RadioButtonGroup name="method" onChange={onChange} >
          <RadioButton value="GET" />
          <RadioButton value="POST" />
          <RadioButton value="PUT" />
          <RadioButton value="PATCH" />
          <RadioButton value="DELETE" />
        </RadioButtonGroup>
        <button>Go!</button>
        <textarea name="jsonBody" value={jsonBody} onChange={onChange}> </textarea>
      </fieldset>
    </form>
  );
};
 
export default Form;
