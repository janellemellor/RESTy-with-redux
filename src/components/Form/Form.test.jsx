import React from 'react';
import { shallow } from 'enzyme';
import Form from './Form';
import { RESTyProvider } from '../../hooks/RESTyProvider';

describe('Form Component', () => {
  it('matches a snapshot', () => {
    const wrapper = shallow(
    <RESTyProvider>
      <Form onSubmit={() => {}} url="https://github.com/janellemellor" onChange={() => {}} jsonBody="{}" />
    </RESTyProvider>);
    expect(wrapper).toMatchSnapshot();
  });
}); 
