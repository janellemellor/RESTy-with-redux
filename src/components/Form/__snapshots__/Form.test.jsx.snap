// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Component matches a snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "jsonBody": "",
      "method": "GET",
      "onChange": [Function],
      "onSubmit": [Function],
      "url": "",
    }
  }
>
  <Form
    jsonBody="{}"
    onChange={[Function]}
    onSubmit={[Function]}
    url="https://github.com/janellemellor"
  />
</ContextProvider>
`;
